### ISSUE TITLE
restrictive class requirements for rowLogSumExps and colLogSumExps 

### ISSUE THREAD

**Issue by StatsMan53** [HenrikBengtsson/matrixStats#132]
rowLogSumExps and colLogSumExps are more restrictive than logSumExp with regard to the class of input values (they object to integer input values, while logSumExp does not).  It would be useful if this restriction were removed.  As an example, consider the following R code:

```r
lse <- matrixStats::logSumExp

mat <- rbind(1:3,4:6)               # 2x3 array, mat == round(mat,0)
matd <- mat + 1e-17                 # perturbed by epsilon, too small to numerically change values

# verify mat and matd contain identical values
mat == matd
matd-mat
# however, R views the contents as different classes
class(mat[1,1])
class(matd[1,1])

# despite being identical, mat and matd produce different results in rowLogSumExps and colLogSumExps, but not in logSumExp
matrixStats::rowLogSumExps(mat)     # why does it refuse to calculate?
matrixStats::rowLogSumExps(matd)    # correct results
c(lse(mat[1,]), lse(mat[2,]))       # logSumExp has no problem

matrixStats::colLogSumExps(mat)     # why does it refuse to calculate?
matrixStats::colLogSumExps(matd)    # correct results
c(lse(mat[,1]), lse(mat[,2]), lse(mat[,3]))
```

**Comment by HenrikBengtsson** [comment]
Thanks. I confirm that can reproduce this:
```r
> lx <- 1:2
> typeof(lx)
[1] "integer"
> matrixStats::logSumExp(lx)
[1] 2.313262

> lx <- matrix(1:2, ncol = 1L)
> typeof(lx)
[1] "integer"
> matrixStats::colLogSumExps(lx)
Error in matrixStats::colLogSumExps(lx) : 
  Argument 'lx' cannot be integer.
```

Although the dominant use case of these functions should be with floating-point values ("double"), for the convenience, I went ahead and [added support for _integers_](https://github.com/HenrikBengtsson/matrixStats/commit/32d5b9793b910fefe1e008effc05f35623040c2a).  I did it by internally pass `lx` to the native code as `as.numeric(lx)`, which means a coercion will take place if and only if `lx` are not already floating-point values.

The updated version is available in the `develop` branch, to be part of the next release.  It can be install as:
```r
> remotes::install_github('HenrikBengtsson/matrixStats@develop')
```

Now, is there really a use case for using _integers_ with `logSumExp()` and friends?  AFAIU, the main purpose of that calculating is improve them numerical accuracy of the result; from `?logSumExp`:
> This function, which avoid numerical underflow, is often used when computing the logarithm of the sum of small numbers (|x| << 1) such as probabilities.
>
> This is function is more accurate than log(sum(exp(lx))) when the values of x = exp(lx) are |x| << 1. [...]

In other words, there should be no advantage of passing _integers_ to this function.  On can equally well use:
```r
> lx <- 1:2
log(sum(exp(lx)))
```
and I suspect one don't even end up having `lx <- log(x)` values in that case, but non-logged `x` values.  So, you can do `log(sum(x))` directly.

PS. The rendering/formatting is [Markdown](https://guides.github.com/features/mastering-markdown/) which supported by lots of sites and frameworks - a very neat format. Some are even moving from LaTeX to Markdown for writing full scientific articles.  In your case, you need to "fence" your code block with `` ``` `` on the lines before and after the code - if you're into LaTeX that corresponds to using `\begin{verbatim}` and `\end{verbatim}`.  I've done that for you comment.
